### PR TITLE
Implement String.prototype.lastIndexOf()

### PR DESCRIPTION
**Pull request by lvidacs**
JerryScript-DCO-1.0-Signed-off-by: Laszlo Vidacs lvidacs.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by lvidacs** [comment]
The helper function is prepared to handle indexOf() as well, it will be adjusted in a separate step.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp [hunk 1]
```diff
@@ -506,6 +515,163 @@ ecma_builtin_helper_string_index_normalize (ecma_number_t index, /**< index */
...
+
+      /* iterate original string and try to match at each position */
+      bool found = false;
+      bool searching = true;
+
+      while (!found && searching)
+      {
+        /* match as long as possible */
+        ecma_length_t match_len = 0;
+        lit_utf8_iterator_pos_t stored_original_pos = lit_utf8_iterator_get_pos (&original_it);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp]: Can't we save the iterator itself? Seeking is not a cheap operation.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h [hunk 2]
```diff
@@ -34,7 +34,11 @@ extern ecma_completion_value_t ecma_builtin_helper_array_concat_value (ecma_obje
                                                                        uint32_t *length,
                                                                        ecma_value_t);
 extern uint32_t ecma_builtin_helper_array_index_normalize (ecma_number_t index, uint32_t length);
-extern uint32_t ecma_builtin_helper_string_index_normalize (ecma_number_t index, uint32_t length);
+extern uint32_t ecma_builtin_helper_string_index_normalize (ecma_number_t index, uint32_t length, bool nan_to_zero);
+extern ecma_completion_value_t ecma_builtin_helper_string_prototype_object_index_of (ecma_value_t this_arg,
+                                                                                            ecma_value_t arg1,
+                                                                                            ecma_value_t arg2,
+                                                                                            bool firstIndex);
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.h]: Incorrect indentation

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp [hunk 3]
```diff
@@ -481,11 +482,19 @@ ecma_builtin_helper_array_concat_value (ecma_object_t *obj_p, /**< array */
  */
 uint32_t
 ecma_builtin_helper_string_index_normalize (ecma_number_t index, /**< index */
-                                            uint32_t length) /**< string's length */
+                                            uint32_t length, /**< string's length */
+                                            bool nan_to_zero) /**< whether NaN is mapped to zero (t) or length (f) */
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp]: This argument modifies the behavior a bit so we need to update the function comment above.

**Comment by lvidacs** [comment]
@galpeter, thanks for the comments, the patch is rebased and updated.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp (hunk 3 above)
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp]: This is the first time this appeared as problem? I think charCodeAt also depends on this.

Comment (lvidacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp]: charCodeAt needs similar checks, but is a bit simpler and does not use the helper function.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp [hunk 4]
```diff
@@ -505,6 +517,163 @@ ecma_builtin_helper_string_index_normalize (ecma_number_t index, /**< index */
...
+               lit_utf8_iterator_read_next (&original_it) == lit_utf8_iterator_read_next (&search_it))
+        {
+          match_len++;
+        }
+
+        /* check for match */
+        if (match_len == search_len)
+        {
+          *ret_num_p = ecma_uint32_to_number (index);
+          found = true;
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp]: Do we need the `found` at all? Can't we just break out?

**Comment by galpeter** [comment]
lgtm

**Comment by zherczeg** [comment]
looks good, but please open an issue that where the followup work is described

**Comment by lvidacs** [comment]
Issue #515 is opened for follow-up work.

**Comment by egavrin** [comment]
`make push`

**Comment by kkristof** [comment]
Landed in https://github.com/Samsung/jerryscript/commit/554305d88c497ff04ef4602ffa93f5b821c7f4af

**Pull request closed without merging**